lib/pure/collections/sequtils.nim

```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
       break
   result
 
-template toSeq1(s: not iterator): untyped =
-  # overload for typed but not iterator
-  type OutType = typeof(items(s))
-  when compiles(s.len):
-    block:
-      evalOnceAs(s2, s, compiles((let _ = s)))
-      var i = 0
-      var result = newSeq[OutType](s2.len)
-      for it in s2:
-        result[i] = it
-        i += 1
-      result
-  else:
-    var result: seq[OutType] = @[]
-    for it in s:
-      result.add(it)
-    result
+template toSeqBuiltin[IX, T](arg: array[IX, T]): seq[T] = @arg
 
-template toSeq2(iter: iterator): untyped =
-  # overload for iterator
-  evalOnceAs(iter2, iter(), false)
-  when compiles(iter2.len):
-    var i = 0
-    var result = newSeq[typeof(iter2)](iter2.len)
-    for x in iter2:
-      result[i] = x
-      inc i
-    result
+template toSeqBuiltin[T](arg: openArray[T]): seq[T] = @arg
+
+template toSeqBuiltin(arg: string): seq[char] = @arg
+
+template toSeqBuiltin[T](arg: seq[T]): seq[T] = arg
+
+proc toSeqBuiltin[T](arg: set[T]): seq[T] {.inline.} =
+  for it in arg:
+    result.add it
+
+proc toSeqBuiltin(a: cstring): seq[char] {.inline.} =
+  let len = a.len
+  result.setLen len
+  var i = 0
+  while i < len:
+    result[i] = a[i]
+    inc i
+
+proc toSeqBuiltin[E: enum](t: typedesc[E]): seq[E] =
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
+  let tmpSym = genSym(nskVar)
+  result = nnkStmtListExpr.newTree
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
+  result.add quote do:
+    for it in `arg`:
+      `tmpSym`.add it
```

Comment:
`:` can be very ambiguous sometimes, so it's better to split it away from command call.

```suggestion
  result.add:
    quote do:
      for it in `arg`:
        `tmpSym`.add it
```

Reply:
Actually I've written a lot of code in this concise way (command call). I think it is very readable. Especially this exact pattern `result.add quote do:` is a pattern I've used a lot in the past. I think my opengl sandbox project is full of it.

Long story short, I'd prefer not not split the line here.

Reply:
It's more about how the compiler sometimes bind the `do:` to the wrong callee, which I tripped on a few times while developing CPS.

It's fine if you left it as-is in this case.